File: src/Day7/Part1.hs
-- Incomplete patterns in this file work for all puzzle inputs
{-# OPTIONS_GHC -Wno-incomplete-patterns #-}

module Day7.Part1 where

import Data.Bits

getFuncs :: Int -> Int -> [(Int -> Int -> Int)]
{-# INLINE getFuncs #-}
getFuncs n ops0 = take n $ step ops0
  where
    step ops = case ops .&. 1 of
      0 -> (+) : step (ops `shiftR` 1)
      1 -> (*) : step (ops `shiftR` 1)

-- Apply binary operations left to right; `length xs0` must be one greater than
-- `length funcs0`.
applyFuncs :: [a -> a -> a] -> [a] -> a
applyFuncs _ [] = error "applyFuncs: No operands provided"
applyFuncs funcs0 (x0 : xs0) =
  foldl (\acc (f, x) -> f acc x) x0 $ zip funcs0 xs0

validateEquation :: (Int, [Int]) -> Bool
validateEquation (testVal, operands) = go 0
  where
    n = length operands - 1
    go ops
      | result == testVal = True
      | popCount ops >= n = False
      | otherwise = go (ops + 1)
      where
        result = applyFuncs (getFuncs n ops) operands

part1 :: [(Int, [Int])] -> Int
part1 equations = sum $ map fst $ filter validateEquation equations

File: rust/src/day7/part1.rs
fn validate_equation(test_val: &u64, operands: &[u64]) -> bool {
    fn go(test_val: u64, operands: &[u64], acc: u64) -> bool {
        match operands {
            [] => acc == test_val,
            [x, ..] => {
                if acc > test_val {
                    false
                } else {
                    go(test_val, &operands[1..], acc + x) || go(test_val, &operands[1..], acc * x)
                }
            },
        }
    }

    if let Some(first) = operands.first() {
        go(*test_val, &operands[1..], *first)
    } else {
        false
    }
}

pub fn part1(equations: &[(u64, Vec<u64>)]) -> u64 {
    equations
        .iter()
        .filter(|(test_val, operands)| validate_equation(test_val, operands))
        .map(|(test_val, _)| test_val)
        .sum()
}
